vault-plugins/shared/memdb/memdb.go

```diff
@@ -0,0 +1,407 @@
+func (t *Txn) checkForeignKeys(table string, obj interface{}, allowedArchivingTimeStamp UnixTime, allowedArchivingHash int64) error {
+	keys := t.schema.MandatoryForeignKeys[table]
+	return t.processRelations(keys, obj, t.checkForeignKey(allowedArchivingTimeStamp, allowedArchivingHash), ErrForeignKey)
+}
+
+// checkForeignKey supports Slice as a field type
+func (t *Txn) checkForeignKey(allowedArchivingTimeStamp UnixTime, allowedArchivingHash int64) func(checkedFieldValue interface{}, key Relation) error {
+	return func(checkedFieldValue interface{}, key Relation) error {
+		switch reflect.TypeOf(checkedFieldValue).Kind() {
+		case reflect.Slice:
+			s := reflect.ValueOf(checkedFieldValue)
+			for i := 0; i < s.Len(); i++ {
+				err := t.checkSingleValueOfForeignKey(s.Index(i).Interface(), key, allowedArchivingTimeStamp, allowedArchivingHash)
+				if err != nil {
+					return err
+				}
+			}
+			return nil
+		default:
+			return t.checkSingleValueOfForeignKey(checkedFieldValue, key, allowedArchivingTimeStamp, allowedArchivingHash)
+		}
+	}
+}
+
+func (t *Txn) checkSingleValueOfForeignKey(singleCheckedFieldValue interface{}, key Relation,
+	allowedArchivingTimeStamp UnixTime, allowedArchivingHash int64) error {
+	relatedRecord, err := t.First(key.RelatedDataType, key.RelatedDataTypeFieldIndexName, singleCheckedFieldValue)
+	if err != nil {
+		return fmt.Errorf("getting related record:%w", err)
+	}
+	if relatedRecord == nil {
+		return fmt.Errorf("FK violation: %q not found at table %q at index %q",
+			singleCheckedFieldValue, key.RelatedDataType, key.RelatedDataTypeFieldIndexName)
+	}
+	a, ok := relatedRecord.(Archivable)
+	if !ok {
+		if allowedArchivingTimeStamp == 0 && allowedArchivingHash == 0 {
+			return nil // related record is not archivable, exists, an no need to check
+		} else {
+			return fmt.Errorf("%w related record %#v is not archivable", ErrNotArchivable, relatedRecord)
+		}
+	}
+	s, h := a.ArchiveMarks()
+	if a.Archived() && (s != allowedArchivingTimeStamp || h != allowedArchivingHash) {
+		return fmt.Errorf("FK violation: %q not found at table %q at index %q",
+			singleCheckedFieldValue, key.RelatedDataType, key.RelatedDataTypeFieldIndexName)
+	}
+	return nil
+}
+
+func (t *Txn) checkCascadeDeletesAndCheckingRelations(table string, obj interface{}) error {
+	rels := append(t.schema.CascadeDeletes[table], t.schema.CheckingRelations[table]...) //nolint:gocritic
+	return t.processRelations(rels, obj, t.checkRelationShouldBeEmpty, ErrNotEmptyRelation)
+}
+
+func (t *Txn) checkCheckingRelations(table string, obj interface{}) error {
+	rels := t.schema.CheckingRelations[table]
+	return t.processRelations(rels, obj, t.checkRelationShouldBeEmpty, ErrNotEmptyRelation)
+}
+
+// implement main loop checking relations
+func (t *Txn) processRelations(relations []Relation, obj interface{},
+	relationHandler func(originObjectFieldValue interface{}, key Relation) error, relationHandlerError error) error {
+	valueIface := reflect.ValueOf(obj)
```

Comment:
The relationHandler type is so long, I didn't parse it right for the first time. Maybe we should give it a name

Reply:
I give a new line for this argument